### PR TITLE
Non-record: 1xH100 Budget Run — SmearGate + BigramHash + MLP3x (1.2774 BPB)

### PR DESCRIPTION
**Pull request by tsubasagit**
## What & Why

Can you compete in Parameter Golf with just **1x H100 and $15**? This submission finds out.

The official leaderboard assumes 8xH100 SXM ($24/h). We adapted the proven techniques from PR #162 (SmearGate, BigramHash, MLP3x, SWA) and tuned hyperparameters specifically for a single-GPU setup on RunPod. The result: **val_bpb 1.2774**, approaching the official naive baseline (1.2244) at 1/8th the hardware.

### 3-Seed Validation

| Seed | val_bpb | val_loss | Artifact (bytes) |
|------|---------|----------|-----------------|
| 1337 | 1.27754 | 2.15706 | 16,374,104 |
| 42 | 1.27402 | 2.15113 | 16,389,057 |
| 7 | 1.28077 | 2.16252 | 16,377,079 |
| **Mean** | **1.27744** | **2.15690** | |
| **Std** | **0.00338** | | |

### Where This Sits

| Submission | Hardware | val_bpb | Cost |
|------------|----------|---------|------|
| SOTA (PR #195) | 8xH100 | 1.1428 | — |
| PR #162 (our base) | 8xH100 | 1.1458 | — |
| Naive Baseline | 8xH100 | 1.2244 | — |
| **This run** | **1xH100** | **1.2774** | **~$15** |

## What We Changed for 1xH100

The core challenge: with 1 GPU you get ~1,250 steps in 10 minutes vs ~7,000+ on 8xH100. Every hyperparameter choice must account for this.

| Parameter | 8xH100 (PR #162) | 1xH100 (ours) | Why |
|-----------|-------------------|---------------|-----|
| train_batch_tokens | 786,432 | 524,288 | Smaller batch = more steps in 10min |
| warmdown_iters | 3,000 | 800 | Proportional to fewer total steps |
| swa_start_frac | 0.5 | 0.7 | Only average well-converged checkpoints |
| swa_every | 50 | 100 | Fewer but higher-quality snapshots |
| train_shards | 80 | 20 | Budget constraint ($15 total) |

Architecture unchanged: 9L/512d, MLP3x, SmearGate, BigramHash(4096), U-Net skip, int6+zlib.

## Lessons Learned

**Data quantity matters more than architecture.** Our first attempt used only 1 training shard and scored 1.5070 — *worse* than the unmodified baseline (1.3599). Increasing to 20 shards improved the score by 0.23 BPB to 1.2774. This single change had 3x more impact than all architecture improvements combined.

**SWA timing needs adjustment for fewer steps.** With only ~1,250 steps, starting SWA at 50% (step ~625) captures under-trained checkpoints. Moving to 70% (step ~875) ensured only converged weights were averaged.

**Cost breakdown:**

| Phase | Time | Cost |
|-------|------|------|
| Data download (20 shards) | 10 min | $0.55 |
| Training x3 seeds | 30 min | $1.65 |
| Sliding window eval x3 | 60 min | $3.30 |
| **Total** | **~2 hours** | **~$5.50** |

## Known Limitations

- **Artifact size exceeds 16MB by ~2.3%.** Using zlib compression; switching to zstd-22 (as in PR #162's original) would likely resolve this (~5% better compression on int6 data).
- **No ablation runs.** Budget constraints prevented isolating individual technique contributions.
- **20 shards vs 80.** More training data would likely improve results further.

## Reproduce

```bash
# On RunPod 1xH100 (or any single H100)
git clone https://github.com/tsubasagit/parameter-golf-1.git
cd parameter-golf-1
git checkout submission/1xh100-budget-run
python3 data/cached_challenge_fineweb.py --variant sp1024 --train-shards 20

RUN_ID=test SEED=1337 torchrun --standalone --nproc_per_node=1 \
  records/track_non_record_16mb/2026-04-08_1xH100_SmearGate_BigramHash_MLP3x/train_gpt.py
```

All defaults in the script are tuned for 1xH100. No environment variables needed beyond RUN_ID and SEED.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: 1xH100 Budget Run — SmearGate + BigramHash + MLP3x (1.2774 BPB)

**BPB:** 1.2774 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `53a60f48b6c0`, file `records/track_non_record_16mb/2026-04-08_1xH100_SmearGate_BigramHash_MLP3x/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.11s, dim=512, layers=9, vocab=1024, code=52243 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.11s, dim=512, layers=9, vocab=1024, code=52243 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by tsubasagit** [comment]
Thanks for the thorough review, @MatoTeziTanka 

  One note on the artifact size: as mentioned in the PR description,
  the current artifacts exceed 16MB by ~2.3% (using zlib). Switching
  to zstd-22 should bring them under the cap — happy to rerun if
  maintainers want that before merge.

  Also, since this is a 1xH100 non-record submission, the ≤600s
  train/eval constraint on 8×H100 may not directly apply — but I can
  clarify timing details if needed.
